### DISCUSSION TITLE
Is it possible to clone/checkout a repo for a given branch or tag?

### DISCUSSION THREAD

**Discussion by lquerel** [GitoxideLabs/gitoxide#1389] [Q&A]
I'm currently using the following code to checkout the head of specific repo into a local directory (in the following example in `git_repo_path`).

```rust
// Clones the repo into the tempdir.
// Use shallow clone to save time and space.
let mut fetch = PrepareFetch::new(
    repo_url.as_str(),
    git_repo_path,
    Kind::WithWorktree,
    create::Options {
        destination_must_be_empty: true,
        fs_capabilities: None,
    },
    open::Options::isolated(),
)
.map_err(|e| GitError {
    repo_url: repo_url.clone(),
    message: e.to_string(),
})?
.with_shallow(Shallow::DepthAtRemote(
    NonZeroU32::new(1).expect("1 is not zero"),
));

let (mut prepare, _outcome) = fetch
    .fetch_then_checkout(progress::Discard, &AtomicBool::new(false))
    .map_err(|e| GitError {
        repo_url: repo_url.clone(),
        message: e.to_string(),
    })?;

let (_repo, _outcome) = prepare
    .main_worktree(progress::Discard, &AtomicBool::new(false))
    .map_err(|e| GitError {
        repo_url: repo_url.clone(),
        message: e.to_string(),
    })?;
```

I am struggling to find the part of the API that allows retrieving the content in the main worktree corresponding to a specific branch or tag.

**Comment by Byron** [comment]
Even though `gitoxide` errors on the side of too-much-complexity-but-everything-is-accessible, in this case I think it's better to use `Repository::prepare_clone()` instead.

Regarding the question, currently there is no API to allow something like `git clone --branch <name>`. However, [here it is hardcoded to `HEAD`](https://github.com/Byron/gitoxide/blob/5f86e6b11bb73921b458ffee9091bc028a7d6204/gix/src/clone/checkout.rs#L91) and it seems quite straightforward to let it check for a configured optional branch name to use instead. I put it [on the roadmap](https://github.com/Byron/gitoxide/issues/470) to not forget, and contributions are welcome.

**Comment by lquerel** [comment]
Thank you. I will use `Repository::prepare_clone()`. 

Regarding the implementation of an equivalent to `git clone --branch <name>`, I need a bit more guidance as I am not familiar with the internal implementation of `gitoxide`. 

Which approach do you recommend following? Should I implement an alternative to `main_worktree` that takes the name of a branch as a parameter and replaces the call to `Repository::head()` with a new method `Repository::branch(name)`? I tried following this approach but without success; it seems that the call to `Repository::find_reference(<branch>)` fails. Any help/instructions would be welcome to speed up the process. Thank you.

**Reply by Byron** [reply]
> Which approach do you recommend following? Should I implement an alternative to `main_worktree` that takes the name of a branch as a parameter and replaces the call to `Repository::head()` with a new method `Repository::branch(name)`? I tried following this approach but without success; it seems that the call to `Repository::find_reference(<branch>)` fails. Any help/instructions would be welcome to speed up the process. Thank you.

That's a great idea, trying it out locally real quick. As for not finding the branch, I don't know, I'd think it should find it after a clone. Note though that it only creates tracking branches and the local branch that `HEAD` points to. This is probably  where some of the difficulty would lie - doing it the way Git does, setting up the correct local branch, testing it… .

As for passing the branch name, you can make it a builder-method on `PrepareCheckout`, then use it if present.

It's OK to cobble something together and submit a draft PR - I could help and/or contribute code there depending on your interest in getting hands-on there.
